sources/api/early-boot-config/src/provider/cdrom.rs

```diff
@@ -0,0 +1,209 @@
+
+        // If no files exist, return
+        let user_data_file = match user_data_files.next() {
+            Some(file) => file,
+            None => return Ok(None),
+        };
+
+        // There should only be 1 file, if more exist something is wonky
+        ensure!(
+            user_data_files.next().is_none(),
+            error::UserDataFileCount {
+                location: Self::CD_ROM_MOUNT
+            }
+        );
+
+        // Read the file; XML files require extra processing.
+        info!("'{}' exists, using it", user_data_file.display());
+        let user_data_str = match user_data_file.extension().and_then(OsStr::to_str) {
+            Some("xml") | Some("XML") => Self::ovf_user_data(user_data_file)?,
+            // Since we only look for a specific list of file names, we should never find a file
+            // with an extension we don't understand.
+            Some(_) => unreachable!(),
+            None => fs::read_to_string(&user_data_file).context(error::InputFileRead {
+                path: user_data_file,
+            })?,
+        };
+
+        // If the file was empty, return
+        if user_data_str.is_empty() {
+            return Ok(None);
+        }
+        trace!("Received user data: {}", user_data_str);
+
+        // Remove outer "settings" layer before sending to API
+        let mut val: toml::Value =
+            toml::from_str(&user_data_str).context(error::TOMLUserDataParse)?;
+        let table = val.as_table_mut().context(error::UserDataNotTomlTable)?;
+        let inner = table
+            .remove("settings")
+            .context(error::UserDataMissingSettings)?;
+
+        let json = SettingsJson::from_val(&inner, "user data").context(error::SettingsToJSON)?;
+        Ok(Some(json))
+    }
+
+    /// Read and base64 decode user data contained in an OVF file
+    // In VMWare, user data is supplied to the host via an XML file.  Within
+    // the XML file, there is a `PropertySection` that contains `Property` elements
+    // with attributes.  User data is base64 encoded inside a `Property` element with
+    // the attribute "user-data".
+    // <Property key="user-data" value="1234abcd"/>
+    fn ovf_user_data<P: AsRef<Path>>(path: P) -> Result<String> {
+        let path = path.as_ref();
+        let file = File::open(path).context(error::InputFileRead { path })?;
+        let reader = BufReader::new(file);
+
+        // Deserialize the OVF file, dropping everything we don't care about
+        let ovf: Environment =
+            serde_xml_rs::from_reader(reader).context(error::XMLDeserialize { path })?;
+
+        // We have seen the keys in the `Property` section be "namespaced" like "oe:key" or
+        // "of:key".  Since we aren't trying to validate the schema beyond the presence of the
+        // elements we care about, we can ignore the namespacing.  `serde_xml_rs` effectively
+        // ignores these namespaces and returns "key" / "value":
```

Comment:
A document with an example of this kind of "namespaced", ideally in a unit test, would help with understanding.

Reply:
Added a few unit tests with sample data.